onnxruntime/python/tools/transformers/fusion_attention.py

```diff
@@ -401,6 +407,10 @@ def fuse(self, normalize_node, input_name_to_nodes, output_name_to_node):
             _, mask_nodes, _ = self.model.match_parent_paths(
                 where_qk, [(['Cast', 'Equal', 'Unsqueeze', 'Unsqueeze'], [0, 0, 0, 0]),
                            (['Equal', 'Unsqueeze', 'Unsqueeze'], [0, 0, 0])], output_name_to_node)
+            if add_qk is not None:
+                gather_elements_nodes = self.model.match_parent_path(add_qk, ['GatherElements', 'Expand', 'Where'], [1, 1, 1], output_name_to_node)
```

Comment:
Those nodes have no impact since add_qk_str is just second input of add_qk. GatherElements, Expand and Where seems have no impact. Also, how could you ensure that add_qk_str has expected 4D shape?

Reply:
good catch, we just need second input.

However, I don't see we verify the shapes of inputs here, it would just be added and later verified with constraints put in inference. Is there a need for it?

Reply:
If possible, use symbolic shape inference result to check whether the input of add_qk meets the shape constraint.

Reply:
updated.